iocore/net/quic/QUICTLS_boringssl.cc

```diff
@@ -40,141 +126,187 @@ QUICTLS::QUICTLS(QUICPacketProtectionKeyInfo &pp_key_info, SSL_CTX *ssl_ctx, Net
     _netvc_context(nvc_ctx)
 {
   ink_assert(this->_netvc_context != NET_VCONNECTION_UNSET);
+
+  if (this->_netvc_context == NET_VCONNECTION_OUT) {
+    SSL_set_connect_state(this->_ssl);
+
+    SSL_set_alpn_protos(this->_ssl, reinterpret_cast<const unsigned char *>(netvc_options.alpn_protos.data()),
+                        netvc_options.alpn_protos.size());
+    const ats_scoped_str &tlsext_host_name = netvc_options.sni_hostname ? netvc_options.sni_hostname : netvc_options.sni_servername;
+    SSL_set_tlsext_host_name(this->_ssl, tlsext_host_name.get());
+  } else {
+    SSL_set_accept_state(this->_ssl);
+  }
+
+  SSL_set_ex_data(this->_ssl, QUIC::ssl_quic_tls_index, this);
+  SSL_set_quic_method(this->_ssl, &quic_method);
+
+  if (session_file && this->_netvc_context == NET_VCONNECTION_OUT) {
+    auto file = BIO_new_file(session_file, "r");
+    if (file == nullptr) {
+      Debug(tag, "Could not read tls session file %s", session_file);
+      return;
+    }
+
+    auto session = PEM_read_bio_SSL_SESSION(file, nullptr, nullptr, nullptr);
+    if (session == nullptr) {
+      Debug(tag, "Could not read tls session file %s", session_file);
+    } else {
+      if (!SSL_set_session(this->_ssl, session)) {
+        Debug(tag, "Session resumption failed : %s", session_file);
+      } else {
+        Debug(tag, "Session resumption success : %s", session_file);
+        this->_is_session_reused = true;
+      }
+      SSL_SESSION_free(session);
+    }
+
+    BIO_free(file);
+  }
 }
 
-int
-QUICTLS::handshake(QUICHandshakeMsgs *out, const QUICHandshakeMsgs *in)
+void
+QUICTLS::set_local_transport_parameters(std::shared_ptr<const QUICTransportParameters> tp)
 {
-  ink_assert(false);
-  return 0;
+  this->_local_transport_parameters = tp;
+
+  uint8_t buf[UINT16_MAX];
+  uint16_t len;
+  this->_local_transport_parameters->store(buf, &len);
+  SSL_set_quic_transport_params(this->_ssl, buf, len);
```

Comment:
>SSL_set_quic_transport_params configures ssl to send params (of length params_len) in the quic_transport_parameters extension in either the ClientHello or EncryptedExtensions handshake message. This extension will only be sent if the TLS version is at least 1.3, and for a server, only if the client sent the extension. The buffer pointed to by params only need be valid for the duration of the call to this function. This function returns 1 on success and 0 on failure.

Just curious when this call failed ?

Reply:
Maybe in case a given ssl context was invalid?
```c
int SSL_set_quic_transport_params(SSL *ssl, const uint8_t *params,                                                                                                                                                                                                                                                                                                     
                                  size_t params_len) {                               
  return ssl->config && ssl->config->quic_transport_params.CopyFrom(                 
                            MakeConstSpan(params, params_len));                      
}
```